File: Source/ChilliSource/Core/Base/Colour.cpp
//
//  Colour.cpp
//  Chilli Source
//  Created by Scott Downie on 27/09/2010.
//
//  The MIT License (MIT)
//
//  Copyright (c) 2010 Tag Games Limited
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

#include <ChilliSource/Core/Base/Colour.h>

#include <ChilliSource/Core/Math/MathUtils.h>

namespace ChilliSource
{
	namespace Core
	{
		const Colour Colour::k_white(1.0f,1.0f,1.0f,1.0f);
		const Colour Colour::k_transparent(0.0f,0.0f,0.0f,0.0f);
		const Colour Colour::k_black(0.0f,0.0f,0.0f,1.0f);
		const Colour Colour::k_grey(0.5f,0.5f,0.5f,1.0f);
		const Colour Colour::k_red(1.0f,0.0f,0.0f,1.0f);
		const Colour Colour::k_green(0.0f,1.0f,0.0f,1.0f);
		const Colour Colour::k_blue(0.0f,0.0f,1.0f,1.0f);
		const Colour Colour::k_magenta(1.0f,0.0f,1.0f,1.0f);
		const Colour Colour::k_yellow(1.0f,1.0f,0.0f,1.0f);
		const Colour Colour::k_cyan(0.0f,1.0f,1.0f,1.0f);
		const Colour Colour::k_cornflowerBlue(0.4f,0.5f,0.8f,1.0f);
		const Colour Colour::k_lightGrey(0.8f,0.8f,0.8f,1.0f);
		const Colour Colour::k_turquoise(0.25f,0.88f,0.82f,1.0f);
		const Colour Colour::k_orange(0.9f,0.4f,0.12f,1.0f);
		const Colour Colour::k_purple(0.5f,0.0f,0.5f,1.0f);
		const Colour Colour::k_pink(0.94f,0.36f,0.47f,1.0f);
		
        //-----------------------------------------------------------
        //-----------------------------------------------------------
        Colour Colour::Clamp(const Colour& in_value, const Colour& in_min, const Colour& in_max)
        {
            Colour result;
            result.r = MathUtils::Clamp(in_value.r, in_min.r, in_max.r);
            result.g = MathUtils::Clamp(in_value.g, in_min.g, in_max.g);
            result.b = MathUtils::Clamp(in_value.b, in_min.b, in_max.b);
            result.a = MathUtils::Clamp(in_value.a, in_min.a, in_max.a);
            return result;
        }
        //-----------------------------------------------------------
		//-----------------------------------------------------------
		Colour::Colour(f32 in_red, f32 in_green, f32 in_blue, f32 in_alpha)
        : r(in_red), g(in_green), b(in_blue), a(in_alpha)
		{
		}
	}
}

File: Projects/Tools/CoreUtils/src/com/chilliworks/chillisource/toolutils/Colour.java
/**
 * Colour.java
 * Chilli Source
 * Created by Ian Copland on 09/10/2014
 * 
 * The MIT License (MIT)
 * 
 * Copyright (c) 2014 Tag Games Limited
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.chilliworks.chillisource.toolutils;

/**
 * A class for containing RGBA colour values in the range 0.0-1.0. After construction the contents of the
 * class are immutable making it safe to pass into methods.
 * 
 * @author Ian Copland
 */
public final class Colour
{
	public static Colour k_white = new Colour(1.0f, 1.0f, 1.0f, 1.0f);
	public static Colour k_black = new Colour(0.0f, 0.0f, 0.0f, 1.0f);
	
	private final float m_r;
	private final float m_g;
	private final float m_b;
	private final float m_a;
	
	/**
	 * Constructs a new colour with the given RGBA values.
	 * 
	 * @author Ian Copland
	 *
	 * @param in_r - The red colour value.
	 * @param in_g - The green colour value.
	 */
	public Colour(float in_r, float in_g, float in_b, float in_a)
	{
		assert (in_r >= 0.0f && in_r <= 1.0f) : "Colour: red value must be in the range 0.0-1.0.";
		assert (in_g >= 0.0f && in_g <= 1.0f) : "Colour: green value must be in the range 0.0-1.0.";
		assert (in_b >= 0.0f && in_b <= 1.0f) : "Colour: blue value must be in the range 0.0-1.0.";
		assert (in_a >= 0.0f && in_a <= 1.0f) : "Colour: alpha value must be in the range 0.0-1.0.";
		
		m_r = in_r;
		m_g = in_g;
		m_b = in_b;
		m_a = in_a;
	}
	/**
	 * @author Ian Copland
	 *
	 * @return The red colour value.
	 */
	public float getR()
	{
		return m_r;
	}
	/**
	 * @author Ian Copland
	 *
	 * @return The green colour value.
	 */
	public float getG()
	{
		return m_g;
	}	
	/**
	 * @author Ian Copland
	 *
	 * @return The blue colour value.
	 */
	public float getB()
	{
		return m_b;
	}
	/**
	 * @author Ian Copland
	 *
	 * @return The alpha value.
	 */
	public float getA()
	{
		return m_a;
	}
	/**
	 * Returns whether or not this object is equal in value to the given
	 * object.
	 * 
	 * @author Ian Copland
	 *
	 * @param in_object - The other object.
	 *
	 * @return Whether or not the given object is equal to this.
	 */
	public boolean equals(Object in_object)
	{
		if (in_object == this)
	    {
	    	return true;
	    }
		
		if (in_object == null)
	    {
	    	return false;
	    }
	    	
	    if ((in_object instanceof Colour) == false)
	    {
	    	return false;
	    }
	    
	    Colour colour = (Colour)in_object;
	    return (Float.compare(m_r, colour.m_r) == 0 && Float.compare(m_g, colour.m_g) == 0 && Float.compare(m_b, colour.m_b) == 0 && Float.compare(m_a, colour.m_a) == 0);
	}
	/**

	 * @author Ian Copland
	 *
	 * @return This object in string form.
	 */
	public String toString()
	{
		return "(" + m_r + ", " + m_g + ", " + m_b + "," + m_a + ")";
	}
}
